### DISCUSSION TITLE
How to start working on parsers for languages using indentation for blocks (Pug, Slim, Slime)

### DISCUSSION THREAD

**Discussion by strayer** [tree-sitter/tree-sitter#1215] [Q&A] [answered]
Hey everyone. I've been using tree-sitter for various languages with nvim 0.5 (nightly) for a while now and it made the editing experience much more pleasant. I'm missing a few languages though that could either benefit *a lot* from tree-sitter or need to work due to multi-language documents. The three languages I'm missing are HTML template engines:

- [Pug](https://pugjs.org/), used in a lot of Vue frontend code I'm working on
- [Slim](http://slim-lang.com/), exclusively used in all of the Rails projects I'm working on
- [Slime](https://slime-lang.com/), pretty much Slim but for Elixir projects

All of these have in common that they are whitespace-sensitive like yaml and Python, so writing a parser for these would require an external scanner. I looked into the scanners for yaml and Python, but find it a bit hard to get a grasp on what is going on (yaml is especially confusing).

Pug and Slim make editing our projects in nvim a bit annoying. Since Pug is used in .vue files mixed with TypeScript and SCSS, I don't have any syntax highlighting _at all_, because the basic vim syntax highlighting obviously can't be used. Slim has a lot of performance problems because the basic vim syntax highlighting imports the full ruby syntax highlighting due to embedded ruby code, resulting in [a lot of pain](https://github.com/slim-template/vim-slim/issues/19). In case of slime I think the implementation of a parser would be mostly identical to slim with the exception of not embedding ruby, but elixir.

The feature set of these languages is very similar, though Pug has more dynamic parts to it than slim(e). I do think that a "basic" syntax highlighting with tree-sitter to make editing a bit more practical would be easily achievable, or at least I hope that one could tackle things bit by bit instead of having to get the whole language with all the specialties working right away.

I'd love to contribute to this, but I'm having trouble finding a place to start. I initially thought I'd just get a very basic HTML-only Pug template working without any embedded languages or conditionals and such:

```pug
doctype html
html
  head
    title MVP
  body
    #some-id
      p.some-class Hello World!
```

After that I could start implementing specific features of pug or slim, step by step, and see where it goes.

But for this the indentation logic in an external scanner would need to work. Can anyone help me get a basic version of this working? I have some very very basic C knowledge from years ago, but especially things like memory management and such are… well… dusty?

**Comment by razzeee** [comment]
Do you have a repo somewhere already?

Also your company website seems broken to me, seems to fail to deliver the css.

**Reply by strayer** [reply]
Hey, thanks for commenting!

> Do you have a repo somewhere already?

Not yet. I started one locally and tried really _really_ simple parsing but quickly realized that I won't be able to continue without figuring out the indentation blocks. I'll get one going later, no harm in having something unfinished on GitHub.

> Also your company website seems broken to me, seems to fail to deliver the css.

Whoops… forwarded that to the people managing the site. 🙈 Thank you!

**Reply by strayer** [reply]
I now got a basic version running based on the Python parser.

```
~/dev/tree-sitter-pug
❯ cat div-test.pug
div
div
  p
    p
  div
div
~/dev/tree-sitter-pug
❯ tree-sitter parse div-test.pug
(document [0, 0] - [6, 0]
  (tag [0, 0] - [0, 3])
  (block [1, 0] - [4, 5]
    (tag [1, 0] - [1, 3])
    (block [2, 2] - [3, 5]
      (tag [2, 2] - [2, 3])
      (tag [3, 4] - [3, 5]))
    (tag [4, 2] - [4, 5]))
  (tag [5, 0] - [5, 3]))
```

Its just a proof of concept for now, I'll maybe put it up as a repo later. It is not really working with deeply nested tags (e.g. the second `p`), but at least the external scanner works. What I need to figure out now is how to structure the rules to make sense for pug. Since it ends up as HTML it probably makes sense to take inspiration from the HTML parser and then figure out how to get the embedded JS and conditionals and so on in there too.

**Comment by ahelwer** [comment, marked as answer]
I recently wrote my own external scanner. I won't lie, it was tricky - the most difficult part (ongoing!) was developing a "theory of mind" about tree-sitter itself and in which circumstances it will call into your external scanner looking for what symbols. The good news is you don't have to handle these cases all at once, off the bat. I had surprising success trying to parse very simple cases and then building up from that, keeping a large battery of tests. Also, if you just make use of things like `std::vector<T>` and other primitive types, you shouldn't have to worry about memory management at all.

With respect to parsing whitespace-sensitive languages, I copied quite a bit from the [Python external scanner](https://github.com/tree-sitter/tree-sitter-python/blob/master/src/scanner.cc). I [eventually found](https://github.com/tree-sitter/tree-sitter/discussions/1160#discussioncomment-891506) that the `INDENT`, `NEWLINE`, and `DEDENT` external tokens worked well for identifying the beginning of the indented block, separation between statements in the same indented block, and the end of the indented block respectively. You can also nest these, so something like:
```
if x > 0:
  x += 1
  if y > 0:
    y += 1
    x = y
  y -= 1
```
Would end up as something like:
```
(if (gt (identifier) (number))
  (indented_block
    (INDENT)
    (plus_equals (identifier) (number))
    (NEWLINE)
    (if (gt (identifier) (number))
      (indented_block
        (INDENT)
        (plus_equals (identifier) (number))
        (NEWLINE)
        (assign (identifier) (identifier))
        (DEDENT)
      )
    )
    (NEWLINE)
    (minus_equals (identifier) (number))
    (DEDENT)
  )
)
```
Where usually you hide the `INDENT`, `NEWLINE`, and `DEDENT` tokens by putting an underscore before them in the externals section of your grammar. Your grammar itself might look like:
```
if: $ => seq('if', $._expr, ':', $.indented_block),
indented_block: $ => seq($.INDENT, repeat(seq($.statement, $.NEWLINE)), $.statement, $.DEDENT),
statement: $ => choice($.plus_equals, $.minus_equals, $.if, $.assign)
```

Your external scanner itself will basically work by keeping track of the current indentation level and its column position with a stack (probably a `std::vector<T>`) that it serializes & deserializes as required. If you lookahead and see that the next statement is in the same column position as the current block, emit a `NEWLINE`. If the next statement is in a column prior to the current block, emit a `DEDENT` then pop the top element of the stack. That sort of thing. Once you have that basic skeleton in place it's easy to implement the actual logic you're looking for.

You can also take a look at [my external scanner](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/main/src/scanner.cc) for a second example.

One final tip - it helps with developing the tree-sitter "theory of mind" to watch tree-sitter's debug output as it parses a file using your external scanner - very illuminating! Use `tree-sitter parse -d filepath`.

**Reply by strayer** [reply]
Thanks for the elaborate comment! I for now simply pulled the python scanner into my WIP pug parser and it seems to work. Needs to be cleaned up, obviously, but its a good base to work on.

As mentioned in the other comment I now need to figure out a structure of rules. Since pug generates HTML templates with a bit of conditionals and embedded code I guess the HTML parser would be a good inspiration, we'll see how far I get with this.

**Reply by zealot128** [reply]
@strayer I've also started forking a [wip repo by another user](https://github.com/RianFuro/tree-sitter-pug) and added multi line Pipes. It works a little bit. I am now struggling to implementing more features like inline JS `` {{ 1 + 1 }}`` as "@javascript". 
That original repo's indent / outdent stuff seemed to work!

https://github.com/zealot128/tree-sitter-pug/blob/master/grammar.js

**Reply by strayer** [reply]
Oh neat, I didn't even see the original repository. Thats way more sophisticated than the stuff I've been experimenting with. I probably won't work on this for a while due to a lot of stuff on my plate, but I'll be checking out your fork later this month for sure.

**Reply by zealot128** [reply]
@strayer I've just added wrapped tags:

```
div.card: div.card-body Content
```

and inline js in content:

```
div
  | foo {{ js here }}
```

It's probably quite unstable but workable for now. (Not multiline js for example). Better than a white unhighlighted code block for now :)

**Reply by strayer** [reply]
Nice! Not sure how you want to go about this. I initially planned to add a TODO block to the README that links to the [pug documentation](https://pugjs.org/api/getting-started.html). One could add a checkbox line for each page in the language reference and implement each page step by step. Since the pages include example code, that could also be used as tests for the tree-sitter parser. Makes it easy to see what still needs to be done.

**Reply by zealot128** [reply]
@strayer Feel free to contribute to the fork. I am done for now and have all the pug syntax that I normally use. I am now struggling to get it working together with the Vue-treesitter as an embedded language.

**Reply by zealot128** [reply]
I will try to take on the Slim parser in the future though. Let's see how much I can reuse from Pug ... :)

**Reply by strayer** [reply]
Hey @zealot128, I'm looking into this again and will try to integrate your tree-sitter fork into my setup. Are you using this with neovim by chance? I'd need to implement queries in nvim-treesitter so it can actually make sense of what is happening. If you already have anything like that working it would probably save a bit of my time fiddling with nvim-treesitter. Depending on how well that goes I might open a PR there to integrate your fork.

**Reply by zealot128** [reply]
I've created a injections.scm into my .config/nvim dir. It's a fork of a file of the same name in tree-sitter-vue with added injection for pug:

```
-- queries/vue/injections.scm

; inherits: html_tags

(
  (style_element
    (start_tag
      (attribute
        (quoted_attribute_value (attribute_value) @_lang)))
    (raw_text) @scss)
  (#any-of? @_lang "scss" "postcss" "less")
)

(
  (script_element
    (start_tag
      (attribute
        (quoted_attribute_value (attribute_value) @_lang)))
    (raw_text) @typescript)
  (#any-of? @_lang "ts" "typescript")
)

((interpolation
  (raw_text) @javascript))

((directive_attribute
    (quoted_attribute_value
      (attribute_value) @javascript)))

(
  (template_element
    (start_tag
      (attribute
        (quoted_attribute_value (attribute_value) @_lang)))
    (raw_text) @pug)
  (#any-of? @_lang "pug")
)

(element
 (start_tag
  (tag_name) @_tag_name
  (attribute
   (attribute_name) @_attr_name
   (quoted_attribute_value
    (attribute_value) @_lang
   )
  )
 )
 (text) @yaml
 (#match? @_tag_name "i18n")
 (#match? @_lang "yaml")
 (#match? @_attr_name "lang")
)


(comment) @comment
```

**Reply by zealot128** [reply]
Also, yes, I am using Nvim with lsp config. To add my Vue + Pug fork to it:

```lua
local parser_config = require("nvim-treesitter.parsers").get_parser_configs()
parser_config.vue.install_info.url = "https://github.com/zealot128/tree-sitter-vue.git"

parser_config.pug = {
   install_info = {
      url = "https://github.com/zealot128/tree-sitter-pug", -- local path or git repo

      files = { "src/parser.c", "src/scanner.cc" },
   },
   filetype = "pug", 
}
```

Then, ``:TSInstallFromSource vue`` + pug (TSInstall or TSInstallFromSource both should work I think).

**Reply by strayer** [reply]
Thanks! I'll try those snippets when I'm working on Vue/Pug again.

**Reply by theoo** [reply]
I'm a bit late to the party. This is the best I could do for now: https://gitlab.com/theoreichel/tree-sitter-slim
